dss/operations/iam.py

```diff
@@ -120,20 +160,112 @@ def dump_fus_policies(fus_client, do_headers):
+    :param client: the boto client to use
+    :param managed: (boolean) if true, include AWS-managed policies
+    :returns: list of tuples of two strings in the form (asset_name, policy_name)
     """
-    pass
+    extracted_list = []
+
+    # Prepare to extract labels for JSON returned by API
+    def _make_aws_api_labels_dict_entry(*args) -> typing.Dict[str, str]:
+        """
+        Convenience function to unpack 4 values into a dictionary of labels, useful for processing
+        API results.
+
+        :params arg[0]: label of asset type
+        :params arg[1]: label of asset name
+        :params arg[2]: label of asset details
+        :params arg[3]: label of asset policy details
+        :returns: dictionary of organized labels
+        """
+        assert len(args) == 4, "Error: need 4 arguments!"
+        return dict(
+            extracted_list_label=args[0], name_label=args[1], detail_list_label=args[2], policy_list_label=args[3]
+        )
+
+    def _get_aws_api_labels_dict() -> typing.Dict[str, typing.Dict[str, str]]:
+        """Store the labels used to unwrap JSON results from the AWS API"""
+        labels = {
+            "user": _make_aws_api_labels_dict_entry("User", "UserName", "UserDetailList", "UserPolicyList"),
+            "group": _make_aws_api_labels_dict_entry("Group", "GroupName", "GroupDetailList", "GroupPolicyList"),
+            "role": _make_aws_api_labels_dict_entry("Role", "RoleName", "RoleDetailList", "RolePolicyList"),
+        }
+        return labels
+
+    # Extract labels needed
+    labels = _get_aws_api_labels_dict()
+    if asset_type not in labels:
+        raise RuntimeError(f"Error: asset type {asset_type} is not valid, try one of: {labels}")
+    extracted_list_label, filter_label, name_label, detail_list_label, policy_list_label = (
+        labels[asset_type]["extracted_list_label"],
+        labels[asset_type]["extracted_list_label"],
+        labels[asset_type]["name_label"],
+        labels[asset_type]["detail_list_label"],
+        labels[asset_type]["policy_list_label"],
+    )
+
+    # Get the response, using paging if necessary
+    response_detail_list: typing.List[typing.Any] = []
+    paginator = client.get_paginator("get_account_authorization_details")
+    for page in paginator.paginate(Filter=[filter_label]):
+        response_detail_list += page[detail_list_label]
+
+    # Inline vs managed policies:
+    # https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_managed-vs-inline.html
+
+    # Order of enumeration:
+    # 1. inline policies
+    # 2. managed policies
+
+    for asset_detail in response_detail_list:
+        asset_name = asset_detail[name_label]
+
+        # 1. Inline policies
+        # Check if any policies are present
+        if policy_list_label in asset_detail:
+            for inline_policy in asset_detail[policy_list_label]:
+                policy_name = inline_policy["PolicyName"]
```

Comment:
should this be able to handle if the policy is missing? or is that case not possible.

Reply:
Line 227, `if policy_list_label in asset_detail`, handles the case where there are no policies attached to the asset; if there are any policies attached to the asset, it will be held under the keys "UserDetailList"/"GroupDetailList"/"RoleDetailList" (which is what `policy_list_label` stores).

Exploring the `get_account_authorization_details` API responses, I did not see any cases where policies were missing names. But I don't know if a policy name is _required_ for a policy to be attached to a user. Will double-check that...

Reply:
From [aws cli docs](https://docs.aws.amazon.com/cli/latest/reference/iam/get-account-authorization-details.html#output) for the `get_account_authorization_details` api endpoint, 

> UserPolicyList -> (list)
    A list of the inline policies embedded in the user.
    (structure)
        Contains information about an IAM policy, including the policy document.
        This data type is used as a response element in the GetAccountAuthorizationDetails operation.
        PolicyName -> (string)
            The name of the policy.
        PolicyDocument -> (string)
            The policy document.

Same is true for GroupPolicyList and RolesPolicyList return values - so we can expect the `PolicyName` field to be present.